### ISSUE TITLE
resample() fails with cryptic error message

### ISSUE THREAD

**Issue by enricoferrero** [mlr-org/mlr#514]
I'm getting this error when using resample() on a large dataset (originally 5000 observations, 25000 features but features should be filtered to 2500 using makeFilterWrapper()).
I'm using R v3.2.0 and mlr v2.4 from CRAN.

`Error in resample(lrn, classif.task, rdesc) :
  Assertion on 'xs' failed: Must be of type 'list', not 'NULL'`

Here's a minimal reproducible example:
```{r}
library(mlr)
library(parallelMap)

set.seed(16)
parallelStart("multicore", 4)

dat <- data.frame(replicate(25000, sample(0:1, 5000, rep=TRUE)))
classif.task <- makeClassifTask(id="test", data=dat, target="X1")
lrn <- makeLearner("classif.svm")
lrn <- makeFilterWrapper(lrn, fw.method="information.gain", fw.perc=0.1)
rdesc <- makeResampleDesc("CV", iters=3)
mod <- resample(lrn, classif.task, rdesc)

parallelStop()
```

Thanks!

**Comment by berndbischl** [comment]
The code you posted runs for me, with CRAN 2.4 and the current 2.5 from GH.

Can you please post sessionInfo and a traceback?

**Comment by berndbischl** [comment]
Also please tell us if the error ONLY occurs if the data set is larg(er). What happens if you reduce it in size?

**Comment by berndbischl** [comment]
Correction: I can reproduce, but I have to wait until the last CV iter. Sorry.

(Although further output and traceback would still have been helpful....)

**Comment by enricoferrero** [comment]
Posting full output:

```{r}
> library(mlr)
Loading required package: BBmisc
Loading required package: ggplot2
library(parallelMap)
Loading required package: ParamHelpers
> library(parallelMap)
> set.seed(16)
> parallelStart("multicore", 2)
Starting parallelization in mode=multicore with cpus=2.
> dat <- data.frame(replicate(25000, sample(0:1, 5000, rep=TRUE)))
classif.task <- makeClassifTask(id="test", data=dat, target="X1")
lrn <- makeLearner("classif.svm")
lrn <- makeFilterWrapper(lrn, fw.method="information.gain", fw.perc=0.1)
rdesc <- makeResampleDesc("CV", iters=3)
mod <- resample(lrn, classif.task, rdesc)
> classif.task <- makeClassifTask(id="test", data=dat, target="X1")
> lrn <- makeLearner("classif.svm")
> lrn <- makeFilterWrapper(lrn, fw.method="information.gain", fw.perc=0.1)
> rdesc <- makeResampleDesc("CV", iters=3)
> mod <- resample(lrn, classif.task, rdesc)
Mapping in parallel: mode = multicore; cpus = 2; elements = 3.
[Resample] cross-validation iter: 1
[Resample] cross-validation iter: 2
[Resample] cross-validation iter: 3
Error in resample(lrn, classif.task, rdesc) :
  Assertion on 'xs' failed: Must be of type 'list', not 'NULL'
> traceback()
9: stop(simpleError(sprintf(msg, ...), call = sys.call(1L)))
8: mstop("Assertion on '%s' failed: %s", var.name, msg)
7: makeAssertion(res, vname(x, .var.name))
6: assertList(xs)
5: extractSubList(iter.results, "measures.train", simplify = "rows")
4: as.data.frame(extractSubList(iter.results, "measures.train",
       simplify = "rows"))
3: mergeResampleResult(learner, task, iter.results, measures, rin,
       models, extract, keep.pred, show.info, runtime)
2: addClasses(mergeResampleResult(learner, task, iter.results, measures,
       rin, models, extract, keep.pred, show.info, runtime), "ResampleResult")
1: resample(lrn, classif.task, rdesc)
> sessionInfo()
R version 3.2.0 (2015-04-16)
Platform: x86_64-unknown-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server release 6.6 (Santiago)

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=C
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=C              LC_PAPER=en_GB.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] parallelMap_1.3      mlr_2.4              ParamHelpers_1.5     ggplot2_1.0.1        BBmisc_1.9           BiocInstaller_1.18.4
[7] vimcom_1.2-6         colorout_1.0-2       setwidth_1.0-4

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.1      magrittr_1.5     splines_3.2.0    MASS_7.3-44      munsell_0.4.2    colorspace_1.2-6 stringr_1.0.0
 [8] plyr_1.8.3       tools_3.2.0      parallel_3.2.0   grid_3.2.0       checkmate_1.6.2  gtable_0.1.2     e1071_1.6-7
[15] class_7.3-14     survival_2.38-3  digest_0.6.8     reshape2_1.4.1   stringi_0.5-5    scales_0.3.0     proto_0.3-10
```

**Comment by enricoferrero** [comment]
I can reproduce it on a significantly smaller dataset:

```{r}
library(mlr)
library(parallelMap)

set.seed(16)
parallelStart("multicore", 4)

dat <- data.frame(replicate(2500, sample(0:1, 500, rep=TRUE)))
classif.task <- makeClassifTask(id="test", data=dat, target="X1")
lrn <- makeLearner("classif.svm")
lrn <- makeFilterWrapper(lrn, fw.method="information.gain", fw.perc=0.1)
rdesc <- makeResampleDesc("CV", iters=3)
mod <- resample(lrn, classif.task, rdesc)

parallelStop()
```

```{r}
> library(mlr)
> library(parallelMap)
> set.seed(16)
> parallelStart("multicore", 4)
Starting parallelization in mode=multicore with cpus=4.
> dat <- data.frame(replicate(2500, sample(0:1, 500, rep=TRUE)))
> classif.task <- makeClassifTask(id="test", data=dat, target="X1")
> lrn <- makeLearner("classif.svm")
> lrn <- makeFilterWrapper(lrn, fw.method="information.gain", fw.perc=0.1)
> rdesc <- makeResampleDesc("CV", iters=3)
> mod <- resample(lrn, classif.task, rdesc)
Mapping in parallel: mode = multicore; cpus = 4; elements = 3.
[Resample] cross-validation iter: 1
[Resample] cross-validation iter: 2
[Resample] cross-validation iter: 3
Error in resample(lrn, classif.task, rdesc) :
  Assertion on 'xs' failed: Must be of type 'list', not 'NULL'
> traceback()
9: stop(simpleError(sprintf(msg, ...), call = sys.call(1L)))
8: mstop("Assertion on '%s' failed: %s", var.name, msg)
7: makeAssertion(res, vname(x, .var.name))
6: assertList(xs)
5: extractSubList(iter.results, "measures.train", simplify = "rows")
4: as.data.frame(extractSubList(iter.results, "measures.train",
       simplify = "rows"))
3: mergeResampleResult(learner, task, iter.results, measures, rin,
       models, extract, keep.pred, show.info, runtime)
2: addClasses(mergeResampleResult(learner, task, iter.results, measures,
       rin, models, extract, keep.pred, show.info, runtime), "ResampleResult")
1: resample(lrn, classif.task, rdesc)
> sessionInfo()
R version 3.2.0 (2015-04-16)
Platform: x86_64-unknown-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server release 6.6 (Santiago)

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=C
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=C              LC_PAPER=en_GB.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] parallelMap_1.3      mlr_2.4              ParamHelpers_1.5     ggplot2_1.0.1        BBmisc_1.9           BiocInstaller_1.18.4
[7] vimcom_1.2-6         colorout_1.0-2       setwidth_1.0-4

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.1      magrittr_1.5     splines_3.2.0    MASS_7.3-44      munsell_0.4.2    colorspace_1.2-6 stringr_1.0.0
 [8] plyr_1.8.3       tools_3.2.0      parallel_3.2.0   grid_3.2.0       checkmate_1.6.2  gtable_0.1.2     e1071_1.6-7
[15] class_7.3-14     survival_2.38-3  digest_0.6.8     reshape2_1.4.1   stringi_0.5-5    scales_0.3.0     proto_0.3-10
```

**Comment by larskotthoff** [comment]
The code you've posted runs fine for me with R 3.2.2. Do you have anything in your `.Rprofile` that may affect this? Could you try with 3.2.2 please?

**Comment by berndbischl** [comment]
Like I said, I can reproduce this on my office laptop. But I need more time to investigate.

**Comment by berndbischl** [comment]
@larskotthoff 
The 1st example really runs for you, until the end, without error?

**Comment by larskotthoff** [comment]
No, I don't have enough memory to run the large example. The smaller one runs fine though.

**Comment by larskotthoff** [comment]
Ok, I can reproduce this for the larger example on a different machine. It doesn't always occur though, I also got

     Mapping in parallel: mode = multicore; cpus = 4; elements = 3.
    [Resample] cross-validation iter: 1
    [Resample] cross-validation iter: 2
    [Resample] cross-validation iter: 3
    Timing stopped at: 16.602 0.261 50.663
    Error : cannot allocate vector of size 2.3 Gb
    Error in stopWithJobErrorMessages(inds, vcapply(result.list[inds], as.character)) :
      Errors occurred in 1 slave jobs, displaying at most 10 of them:

    00001: Error : cannot allocate vector of size 2.3 Gb

**Comment by berndbischl** [comment]
I was guessing before that this would be an out-of-memory error, where we simply see a strange error message, which is probably nothing that can be fixed in mlr.

**Comment by berndbischl** [comment]
In you case this is visible, and "ok" as one sees the problem.

**Comment by larskotthoff** [comment]
Looks like this is an issue with `parallelMap` more than mlr. If I use socket parallelization, I get

    Exporting objects to slaves for mode socket: .mlr.slave.options
    Mapping in parallel: mode = socket; cpus = 4; elements = 3.
    Error in unserialize(node$con) : error reading from connection

I.e. it looks like something just dies and `parallelMap` returns `NULL` instead of reporting an error.

**Comment by berndbischl** [comment]
I am now on my home laptop (which is pretty similar to my office laptop....)
The 2nd example with the smaller data set runs, without problems.
The 1st freezes my laptop for about a min,  no mouse movement possible. 
After this, the 3 processes are gone in top. I would guess at least one of them was killed. 
I am now waiting for about 5 min, but the master process does not seem to return...

**Comment by berndbischl** [comment]
I can now produce a 
Segmentation fault (core dumped)

I changed this:
- no parallelStart, so sequential
- 500 features instead of 5000
- rpart instead of svm

This is really NOT us, doing this here. I am betting on FSelector in combo with Weka and Java.

**Comment by larskotthoff** [comment]
I'm getting various combinations of segmentation faults and

    Error: protect(): protection stack overflow

depending on whether I leave the filter wrapper in or not, use resampling or just a single call to `train`, calling the methods directly or through mlr.

Bottom line: you're running out of memory and all bets are off. We can't really catch this in mlr.

**Comment by berndbischl** [comment]
I am really not sure if this is always an out-of-mem error. For reference:

1) This segfaults in FSelector (or below, in RWeka)
```
set.seed(16)
dat <- data.frame(replicate(25000, sample(0:1, 500, rep=TRUE)))
z = FSelector::information.gain(X1 ~ ., data = dat)
```

2)  "error: protect(): protection stack overflow"
library(e1071)
set.seed(16)
dat <- data.frame(replicate(25000, sample(0:1, 5000, rep=TRUE)))
svm(X1 ~ ., data = dat)

I see zero chance or reason to follow up on this further in mlr. BOTH of these things should never happen. 

**Please file error reports with the maintainers of these packages. We are happy to look at this again when this is fixed upstream.**

A further hint: 1) seems to be your showstopper here, the svm-error we probably just discovered by chance. I guess this is related to the high number of features, which you filter down in your example. mlr offers a couple of other filters, which might not have this problem, so you could maybe just try to change the method for now.

I have also at some time with @mllg looked at other, faster, better options to filter features, but we were never able to finalize our code. IMHO better options exists in other toolkits, somebody just needs to connect them to R. And I currently lack time to do this.

**Comment by enricoferrero** [comment]
Thanks both.
@berndbischl  Yes, it looks like that error is only generated when doing things in parallel, otherwise R just segfaults.

Failure does seem to be related to the filter method. Both chi.squared and information.gain from FSelector fail, while anova.test, kruskal.test and variance do work. 

I have another issue with univariate which I'll report separately.

**Comment by berndbischl** [comment]
@enricoferrero 
> Yes, it looks like that error is only generated when doing things in parallel, otherwise R just segfaults.

Well, not exactly sure what that means. I mean, a segfault (here without parallel) is pretty bad. This could happen (and would guess it does) in parallel on the slave too. Maybe we just dont see the correct error msg, but something else instead. The problem is that this is not some kind of an exception, but a very bad kind if crash, and a) I have zero options to handle this as I lose complete control b) and we are now in the land of "undefined" behavior. I think this should really be filed against FSelector!

Funnily @larskotthoff is the unfortunate maintainer of this package :)
Seriously Lars, can this at least be reported? 

> Both chi.squared and information.gain from FSelector fail, 
> while anova.test, kruskal.test and variance do work.

Sure, this I why I mentioned this, the code is from different packages, there is no reason why other filter code chould fail.

Also: Thx for the dilligent reports.

**Comment by larskotthoff** [comment]
Sure you can report it: https://github.com/larskotthoff/fselector

FSelector heavily uses RWeka though and doesn't do anything that fancy on top of it, so the problem is almost certainly in RWeka. I'll have a look at it when I get a chance, but similar to mlr I don't think there's a lot that can be done in FSelector.

**Comment by larskotthoff** [comment]
Ok, it's not even RWeka -- it's `model.frame`:

    dat <- data.frame(replicate(25000, sample(0:1, 5000, rep=TRUE)))
    model.frame(X1~., dat, na.action = NULL)

This gives me "killed" and when run inside FSelector a segmentation fault. Nothing at all I can do about this in the package.

**Comment by mllg** [comment]
Have you tried to start R with the the argument `--max-ppsize=100000`?

**Comment by larskotthoff** [comment]
This will help with the `protect()` issue, but not with the memory issues.

**Comment by berndbischl** [comment]
The WEKA bug is the "worse" one here, yes. And not a protect stack thing. 
Lars, how do you really know that this is a mem issue?
Wouldnt java tell us this, if this would be the case?

**Comment by berndbischl** [comment]
Sorry, I did not see your post about model.frame. I need to look at this closer.... This looks strange...

**Comment by berndbischl** [comment]
With your modelframe example code I get a protect stack overflow.

**Comment by berndbischl** [comment]
Even if I reduce to 500 columns. Am I stupid, how can this be? 25K x 500 seems to be a rather normal size....?

**Comment by berndbischl** [comment]
A related post, although no help
http://stackoverflow.com/questions/27677391/r-stack-overflow-error-with-randomforest-on-large-dataset-48-512-gb-ram

**Comment by larskotthoff** [comment]
Looks like a bug in R to me actually -- something is allocating memory and not checking the return value.

**Comment by berndbischl** [comment]
Yup, likely.

**Comment by enricoferrero** [comment]
Was the bug reported somewhere upstream?

**Comment by enricoferrero** [comment]
Just checking if the bug was reported upstream? Happy to do it myself but I'm not sure to what project I should report to and how? Cheers,

**Comment by larskotthoff** [comment]
I didn't report it, so please feel free to do that.

**Comment by tjl3** [comment]
Hello. I have been experiencing similar issues when using information.gain but not anova.test (thus, it seems to be related to the FSelector package)

R 3.3.0
mlr 2.8 (release)

Mac OSX Yosemite:
The following error is thrown - Error: protect(): protection stack overflow
I can prevent this error by using the --pp-size flag, but then it will start using too much memory for my machine 6+GB and keeps rising.

Linux (Red-Hat) 
Segmentation fault

In any case, just checking to see if that bug got reported upstream since the last post, and if not, where I should send it.

**Comment by larskotthoff** [comment]
It sounds like this is a memory issue. There's nothing we can do about it in the code I'm afraid.

**Comment by enricoferrero** [comment]
@tjl3 I haven't got around reporting it in the end.

**Comment by tjl3** [comment]
I probably ought to have mentioned this before, but I am using genetic datasets with 20,000 columns (features) and <1000 rows. I did not notice these issues when using a smaller dataset with <1000 columns and <1000 rows.

**Comment by larskotthoff** [comment]
What works and what doesn't depends more on the learners/preprocessing methods that you're using than on mlr. All of the core ML methods are implemented in other packages. mlr itself shouldn't add any additional restrictions in terms of memory in most cases.

**Comment by berndbischl** [comment]
+1 to what Lars  said. 
@tjl3 have you read the complete thread? I thought we pretty exactly pinned down the source of the problem. And it does not seem to be code under our control? 

(If you know feature filter implementation in R that scale better we can talk about interfacing those)

**Comment by tjl3** [comment]
Thank you for the information and for working on this project!

**Comment by schiffner** [comment]
> (If you know feature filter implementation in R that scale better we can talk about interfacing those)

If they scale better needs to be determined, but in #878 CORElearn was suggested.
